src/celengine/body.cpp

```diff
@@ -1075,48 +1054,46 @@ Body::addReferenceMark(ReferenceMark* refMark)
 void
 Body::removeReferenceMark(const string& tag)
 {
-    if (referenceMarks != nullptr)
-    {
-        ReferenceMark* refMark = findReferenceMark(tag);
-        if (refMark)
-        {
-            referenceMarks->remove(refMark);
-            delete refMark;
-            recomputeCullingRadius();
-        }
-    }
+    if (referenceMarks == nullptr)
+        return;
+
+    auto iter = std::find_if(referenceMarks->begin(),
+                             referenceMarks->end(),
+                             [&tag](const auto& rm) { return rm->getTag() == tag; });
+    if (iter == referenceMarks->end())
+        return;
+
+    referenceMarks->erase(iter);
+    recomputeCullingRadius();
 }
 
 
 /*! Find the first reference mark with the specified tag. If the body has
  *  no reference marks with the specified tag, this method will return
  *  nullptr.
  */
-ReferenceMark*
+const ReferenceMark*
 Body::findReferenceMark(const string& tag) const
 {
     if (referenceMarks == nullptr)
         return nullptr;
 
-    if (auto iter = std::find_if(referenceMarks->begin(),
-                                 referenceMarks->end(),
-                                 [&tag](const auto* rm) { return rm->getTag() == tag; });
-        iter != referenceMarks->end())
-    {
-            return *iter;
-    }
-
-    return nullptr;
+    auto iter = std::find_if(referenceMarks->begin(),
+                             referenceMarks->end(),
+                             [&tag](const auto& rm) { return rm->getTag() == tag; });
+    return iter == referenceMarks->end()
+        ? nullptr
+        : iter->get();
 }
 
 
 /*! Get the list of reference marks associated with this body. May return
  *  nullptr if there are no reference marks.
  */
-const list<ReferenceMark*>*
+const std::list<std::unique_ptr<ReferenceMark>>*
```

Comment:
exposing internal implementation sounds like an anti-pattern. perhaps it's better to to return a  new list containing raw pointers

Reply:
Another possibility I was considering here is to bring in range-v3 and expose this as a transform view on the inner collection. The actual type name would probably still expose the implementation details, but the consuming code would not have to interact with the unique_ptr. (And the dependency could go away in the distant future when we can upgrade to C++20)

Reply:
Actually scratch that, ranges-v3 does not build with exceptions disabled, and the author has stated he is not interested in supporting non-standard C++ dialects.

Reply:
Added a transform view implementation. Bonus is that the `getAlternateSurfaceNames` method no longer needs to allocate a vector every time it is called.